airbyte-integrations/connectors/source-mongodb/lib/mongodb_configured_stream/incremental.rb

```diff
@@ -0,0 +1,82 @@
+require_relative '../airbyte_logger.rb'
+
+require_relative './base.rb'
+require_relative '../mongodb_types_converter.rb'
+
+class MongodbConfiguredStream::Incremental < MongodbConfiguredStream::Base
+  DATETIME_FIELD_PARTS = %w{time date _at timestamp ts}
+  CURSOR_TYPES = {
+    datetime: 'DATETIME',
+    integer: 'integer',
+  }
+
+  def new(configured_stream:, state:)
+    super
+
+    value = @state.get(stream_name: stream_name, cursor_field: cursor_field)
+    @cursor =  value && convert_cursor(value)
+  end
+
+  def cursor_field
+    @cursor_field ||= configured_stream['cursor_field']&.first
+  end
+
+  def compose_query
+    if @cursor
+      {
+        cursor_field => {
+          "$gte": @cursor
+        }
+      }
+    else
+      {}
+    end
+  end
+
+  def valid?
+    if configured_stream['cursor_field'].count != 1
+      AirbyteLogger.log("Stream #{stream_name} has invalid configuration. Cursor field #{wrapper['cursor_field']} configuration is invalid. Should contain exactly one document property name.", Level::Fatal)
+      return false
+    end
+
+    true
+  end
+
+  def after_item_processed(item)
+    super
+
+    converted_cursor = convert_cursor(item[cursor_field])
+    if !@cursor || converted_cursor && converted_cursor > @cursor
+      @cursor = converted_cursor
+    end
+  end
+
+  def after_stream_processed
+    super
+
+    @state.set(stream_name: stream_name, cursor_field: cursor_field, cursor: @cursor)
+    @state.dump_state!
+  end
+
+  private
+
+  # Rely on a descriptive naming for type definition. It's too expensive to check the cursor type of every document in advance.
```

Comment:
What do you think about that solution? I do not see any other inexpensive way to determine the cursor type.

Reply:
what if we do something like what we do in [`discover_property_type`](https://github.com/airbytehq/airbyte/pull/2125/files#diff-f28873de3c945904b586e78fa4a1bb1214aa57425680b79206f359984d49734cR67) when the stream is initialized instead? This way we only make a query for say, 1k documents, and determine the type based on that.